### ISSUE TITLE
Openssl 3.0.5 support both provider and engine

### ISSUE THREAD

**Issue by hustcathy** [openssl/openssl#20726]
Hi, I'm recently upgrading openssl from 1.1 to 3.0.5, and after I built it with no-module to support some legacy provider, I found the code of loading engine failed. I know it probably be caused by this:
no-module
Don't build any dynamically loadable engines.

This also implies no-dynamic-engine.

Is there any way to support both loading provider and the engine?

**Comment by hustcathy** [comment]
I'm using this piece of code to load engine:
```
ENGINE_load_dynamic();
ENGINE* Engine = ENGINE_by_id("myengineid");
ENGINE_init(Engine);
```

**Comment by levitte** [comment]
The obvious answer is: don't configure with `no-module`

That being said, I'm not sure I understood why you configured with `no-module`, perhaps there's a nuance that I'm not seeing?

**Comment by hustcathy** [comment]
My code also has some other usage to parse PKCS12 file which should use legacy APIs, so I have to build the openssl with no-module to load the legacy provider.

**Comment by paulidale** [comment]
`no-module` is not required if you want to load the legacy provider.

**Comment by t8m** [comment]
@hustcathy just build without no-module and install the legacy.dll that is built into a path where the OpenSSL library expects it to be loaded from. Or you can also override that path.

Also it is always a good idea to use the latest stable version from your preferred branch (i.e., 3.0.8 or 3.1.0).

**Comment by Feelnblessed** [comment]
Ok
That's cute!
Got it, spammed up,
My bad!

Love ya'!

Mad respect!


Now go get me fucking coffee!



Genius at work here!
Ja!

JFK

CHAD ROLANDO PIERRE RODRIGUEZ
+15628526170
ALPHABET INC (GOOGLE)
1600 AMPHITHEATRE PARKWAY
MOUNTAIN VIEW, CA. 94043

**Comment by hustcathy** [comment]
@t8m , I have tried but don't know how to generate the legacy.so in my linux environment, could you give me some advice on how to build and generate legacy.so?

**Comment by paulidale** [comment]
Configure without `no-module`.  Run `make`.  `legacy.so` should now be in the `providers` directory.

**Comment by hustcathy** [comment]
Thanks. @paulidale .
"./Config and make" can generate the providers/legacy.so now.

**Comment by hustcathy** [comment]
I now have problem in loading the legacy provider, I prebuild the openssl and all the required libs are under $mypath. When running below code:
```
  OSSL_PROVIDER* legacy;
  OSSL_PROVIDER* deflt;
  /* Load Multiple providers into the default (NULL) library context */
  legacy = OSSL_PROVIDER_load(NULL, "legacy");
  if (legacy == NULL) {
    printf("Failed to load Legacy provider\n");
    ERR_print_errors_fp(stdout);
    exit(EXIT_FAILURE);
  }
  deflt = OSSL_PROVIDER_load(NULL, "default");
  if (deflt == NULL) {
    printf("Failed to load Default provider\n");
    OSSL_PROVIDER_unload(legacy);
    exit(EXIT_FAILURE);
  }
```
and I got errors 
```
Failed to load Legacy provider
406721CD7B7F0000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(/extra/xxx/yyy/ssl-55977/dist/host/3pp/lib/legacy.so): /extra/xxx/yyy/ssl-55977/dist/host/3pp/lib/legacy.so: cannot open shared object file: No such file or directory
406721CD7B7F0000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:160:
406721CD7B7F0000:error:078C0105:common libcrypto routines:provider_init:init fail:crypto/provider_core.c:946:name=legacy
```
However, I checked the path "/extra/xxx/yyy/ssl-55977/dist/host/3pp/lib/legacy.so" and it does exist.
Can anyone help on this problem? Thanks.

**Comment by t8m** [comment]
Could you try using strace to see what files are being really loaded?

**Comment by hustcathy** [comment]
How to use strace in the program?
I run the command line using "OPENSSL_MODULES=xxx/lib ./openssl pkcs12 -info -in file.p12 -nodes -legacy", and it runs successfully, the legacy.so is under xxx/lib, which means the libraries are OK.
However, then using code to load the legacy.so below with enviroment variable set:
```
setenv("OPENSSL_MODULES", "xxx/lib/", 1);
OSSL_PROVIDER* legacy;
/* Load Multiple providers into the default (NULL) library context */
legacy = OSSL_PROVIDER_load(NULL, "legacy");
if (legacy == NULL) {
    printf("Failed to load Legacy provider\n");
    ERR_print_errors_fp(stdout);
    exit(EXIT_FAILURE);
}
```
the ERR_print_errors_fp reported that 
```
40364CBDEF2A0000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:119:filename(xxx/lib/legacy.so): xxx/lib/legacy.so: cannot open shared object file: No such file or directory
cathy: in init 0, opts: 262144
40364CBDEF2A0000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:160:
cathy: in init 0, opts: 262144
40364CBDEF2A0000:error:0788000A:common libcrypto routines:provider_init:DSA lib:crypto/provider_core.c:933:name=legacy
```
Any other way to print more debug trace? @t8m.

**Comment by beldmit** [comment]
`OPENSSL_MODULES=xxx/lib strace - o log ./openssl pkcs12 -info -in file.p12 -nodes -legacy` should do the trick

**Comment by hustcathy** [comment]
The command line works well when loading legacy provider and use it to parse legacy pkcs12 file, but I still paste the result of running the command line with strace here:
```
-bash-4.2$ OPENSSL_MODULES=/extra/cwei1/branches/ssl-55977/dist/host/3pp/lib/ strace -o log ../../../dist/host/3pp/bin/openssl pkcs12 -info -in ../../cpp/crypto/unittests/packaging-old.p12 -nodes -legacy
Enter Import Password:
MAC: sha1, Iteration 2048
MAC length: 20, salt length: 8

PKCS7 Encrypted data: xxxxx, Iteration 2048

Certificate bag
Bag Attributes
    localKeyID: xxxxx
Cert content
PKCS7 Data
Shrouded Keybag: xxxx, Iteration 2048
Bag Attributes
    localKeyID: xxxx
Key Attributes: <No Attributes>
KEY content
```
My c++ program couldn't load the provider, anything I can do to enable the strace in my c++ program?

**Comment by hustcathy** [comment]
I found that when running c++ code, it will call ossl_no_config_int(); while using command line, it calls ossl_config_int(); Not sure if this leads to the final different result.
@t8m , when loading legacy.so dynamically, do I have to use openssl.cnf to load the provider?

**Comment by t8m** [comment]
No, you should not need a config file to load the legacy provider. It is needed only for loading the fips provider.

**Comment by t8m** [comment]
When executing the c++ program you can call it from strace the same way as you call the `openssl pkcs12` command.

**Comment by hustcathy** [comment]
Hi @t8m . Thanks for your help.
Finally I know what happened, my c++ program runs in docker environment, and the legacy.so is in my host PC, the docker environment doesn't know the path of the host PC.

**Comment by hustcathy** [comment]
@t8m , how does openssl 3.0 select ciphers when both default and legacy providers are loaded? Try one provider and then another?

**Comment by t8m** [comment]
The order is not defined. It basically fetches the cipher implementation from the provider where it finds it in first.

**Comment by t8m** [comment]
But please note that legacy and default providers support completely different set of ciphers with one exception - the RIPEMD160 hash which is present in both for historic reasons.
